controllers/flow_tree.go

```diff
@@ -0,0 +1,326 @@
+	ytv1.UpdateStateWaitingForOpArchiveUpdatingPrepare: func(ctx context.Context, ytsaurus *apiProxy.Ytsaurus, componentManager *ComponentManager) stepResultMark {
+		if !componentManager.needSchedulerUpdate() || ytsaurus.IsUpdateStatusConditionTrue(consts.ConditionNotNecessaryToUpdateOpArchive) {
+			return stepResultMarkHappy
+		}
+		if ytsaurus.IsUpdateStatusConditionTrue(consts.ConditionOpArchivePreparedForUpdating) {
+			return stepResultMarkUnhappy
+		}
+		return stepResultMarkUnsatisfied
+	},
+	ytv1.UpdateStateWaitingForOpArchiveUpdate: flowCheckStatusCondition(consts.ConditionOpArchiveUpdated),
+	ytv1.UpdateStateWaitingForQTStateUpdatingPrepare: func(ctx context.Context, ytsaurus *apiProxy.Ytsaurus, componentManager *ComponentManager) stepResultMark {
+		if !componentManager.needQueryTrackerUpdate() {
+			return stepResultMarkHappy
+		}
+		if ytsaurus.IsUpdateStatusConditionTrue(consts.ConditionQTStatePreparedForUpdating) {
+			return stepResultMarkUnhappy
+		}
+		return stepResultMarkUnsatisfied
+	},
+	ytv1.UpdateStateWaitingForQTStateUpdate: flowCheckStatusCondition(consts.ConditionQTStateUpdated),
+	ytv1.UpdateStateWaitingForYqlaUpdatingPrepare: func(ctx context.Context, ytsaurus *apiProxy.Ytsaurus, componentManager *ComponentManager) stepResultMark {
+		if !componentManager.needYqlAgentUpdate() {
+			return stepResultMarkHappy
+		}
+		if ytsaurus.IsUpdateStatusConditionTrue(consts.ConditionYqlaPreparedForUpdating) {
+			return stepResultMarkUnhappy
+		}
+		return stepResultMarkUnsatisfied
+	},
+	ytv1.UpdateStateWaitingForYqlaUpdate:       flowCheckStatusCondition(consts.ConditionYqlaUpdated),
+	ytv1.UpdateStateWaitingForSafeModeDisabled: flowCheckStatusCondition(consts.ConditionSafeModeDisabled),
+
+	ytv1.UpdateStateWaitingForPodsRemoval: func(ctx context.Context, ytsaurus *apiProxy.Ytsaurus, componentManager *ComponentManager) stepResultMark {
+		if componentManager.arePodsRemoved() {
+			return stepResultMarkHappy
+		}
+		return stepResultMarkUnsatisfied
+	},
+	ytv1.UpdateStateWaitingForPodsCreation: func(ctx context.Context, ytsaurus *apiProxy.Ytsaurus, componentManager *ComponentManager) stepResultMark {
+		if componentManager.allReadyOrUpdating() {
+			return stepResultMarkHappy
+		}
+		return stepResultMarkUnsatisfied
+	},
+}
+
+func buildFlowTree(components []ytv1.Component) *flowTree {
+	st := newSimpleStep
+	head := st(ytv1.UpdateStateNone)
+	tree := newFlowTree(head)
+
+	updMasterOrTablet := hasMaster(components) || hasTablets(components)
+	updMaster := hasMaster(components)
+	updTablet := hasMaster(components)
+	updStateless := hasStateless(components)
+
+	// TODO: if validation conditions can be not mentioned here or needed
+	tree.chain(
+		newConditionalForkStep(
+			ytv1.UpdateStatePossibilityCheck,
+			// This is the unhappy path.
+			st(ytv1.UpdateStateImpossibleToStart),
+			// Happy path will be chained automatically.
+		),
+	).chainIf(
```

Comment:
The chaining approach allows us to not think about what state will be next when adding it, which simplifies things comparing to the original PR.